### PR TITLE
Numerical Aggregate Fix

### PR DESCRIPTION
**Pull request by haikalpribadi**
# Why is this PR needed?

What we mean by Numerical Aggregates are `sum`, `min`, `max`, `mean`, `median` and `std`.
1. These aggregate functions return 0 when there are no attributes to apply the calculation over. This is wrong, it should return null.
2. These aggregate functions (except for `sum`) returns optional, which does not work with GRPC (before and after the new GRPC refactor).

# What does the PR do?

1. Removed `Optional<T>` return types from aggregate functions.
2. Modify the aggregate functions to return `null` when there are no attributes to calculate over.
3. Fix `RemoteQueryExecutor.run(AggregateQuery)` to consider `null` results.
4. Add tests for `null` results in `AggregateTest`
5. Add tests for numerical aggregate functions in `ServerRPCIT`

Additionally, we also did some cleaning:
1. Removed the first Generics parameter `T` in `Aggregate<T, S>`.
2. Added tests for ordering `String`s and `Date`s in `MatchTest.java`

# Does it break backwards compatibility?

Yes. Aggregate `min` and `max` no longer apply to `String` and `Date`. This is because making the change allows us to simplify numerical aggregates, while still being able to get the same functionality of `min/max string/date` through `order string/date asc/desc`.

# List of future improvements not on this PR

This PR will be followed by another PR where we implement a proper GRPC message for Aggregate responses, and also fix the Remote `GroupAggregate` query which is currently broken too on the client side.

### CODE REVIEW

**Comment by haikalpribadi** [comment]
I don't know why Circle-CI is not showing all the tests on this page. But we can see them here: https://circleci.com/gh/haikalpribadi/grakn/tree/aggregate-numeric-fix

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/aggregate/MinAggregate.java [hunk 1]
```diff
@@ -39,25 +36,29 @@
     }
 
     @Override
-    public Optional<T> apply(Stream<? extends Answer> stream) {
-        return stream.map(this::getValue).min(naturalOrder());
+    public Number apply(Stream<? extends Answer> stream) {
+        NumberPrimitiveTypeComparator comparator = new NumberPrimitiveTypeComparator();
```
Comment (jmsfltchr) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/aggregate/MinAggregate.java]: Design not ideal, but probably a non-issue

Comment (jmsfltchr) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/aggregate/MinAggregate.java]: Ideally the comparator instance would be passed in via constructor, however in this case this may make sense since the comparator implementation is unlikely to need to be swapped out for something else with the same interface in the future. Let's discuss!

File: client-java/src/main/java/ai/grakn/client/executor/RemoteQueryExecutor.java [hunk 2]
```diff
@@ -76,7 +77,9 @@ public void run(UndefineQuery query) {
 
     @Override
     public <T> T run(AggregateQuery<T> query) {
-        return (T) Iterators.getOnlyElement(tx.query(query));
+        Iterator iterator = tx.query(query);
```
Comment (kasper-piskorski) [inline client-java/src/main/java/ai/grakn/client/executor/RemoteQueryExecutor.java]: is it better than not throwing exception?
